index-template.md

```diff
@@ -0,0 +1,94 @@
+# OCI Index Template Protocol
+
+This is version 1 of this specification.
+
+The index-template protocol is configured via a single [URI Template][rfc6570].
+When configured via a [`refEngines` entry](ref-engine-discovery.md#ref-engines-objects), the `uri` property MUST be set, and its value is the URI Template.
+
+Consumers MUST provide at least the following variables:
+
+* `name`, matching `host-based-image-name` in the [`host-based-image-name` rule](host-based-image-names.md).
+* `host`, matching `host` in the `host-based-image-name` rule.
+* `path`, matching `path-rootless` in the `host-based-image-name` rule.
+* `fragment`, matching `fragment` in the `host-based-image-name` rule.
+    If `fragment` was not provided in the image name, it defaults to an empty string.
+
+and expand the URI Template as defined in [RFC 6570 section 3][rfc6570-s3].
+
+The server providing the expanded URI MUST support requests for media type [`application/vnd.oci.image.index.v1+json`][index].
+Servers MAY support other media types using HTTP content negotiation, as described in [RFC 7231 section 3.4][rfc7231-s3.4] (which is [also supported over HTTP/2][rfc7540-s8]).
+
```

Comment:
I think the object data SHOULD be built and stored in discovery service permanently, while discovery service will synchronize the object content. So Consumers retrieve this object from discovery backend directly, that approach I feel is more like what we discussed before.

Your language `Consumers retrieving application...` seems to be the workflow:
1. client request to discovery URI;
2. discovery service retrieve newest descriptor from distribution URI;
3. discovery service package `descriptor` and `casEngines` insidely;
4. discovery service returns the packaged object to client;

Reply:
> I think the object data SHOULD be built and stored in discovery service permanently, while discovery service will synchronize the object content. So Consumers retrieve this object from discovery backend directly, that approach I feel is more like what we discussed before.
>
> Your language `Consumers retrieving application...` seems to be the workflow:
>
> 1. client request to discovery URI;
> 2. discovery service retrieve newest descriptor from distribution URI;
> 3. discovery service package `descriptor` and `casEngines` insidely;
> 4. discovery service returns the packaged object to client;

I'd use “ref engine” where you use “discovery service”, but yeah, having a proxy ref engine that resolves a name via a second ref engine and injects `casEngines` entries is possible (and using such a proxy is one way around the [“both my ref-engine discovery and terminal ref-engine service are dumb” issue][1].  But with [`casEngines` added to image-spec as a descriptor property][2], folks who wanted to could certainly set it statically in the `application/vnd.oci.image.index.v1+json`, so you can have a completely static ref-engine service if you wanted.

[1]: https://github.com/xiekeyang/oci-discovery/issues/1#issuecomment-327639553
[2]: https://github.com/xiekeyang/oci-discovery/pull/2#discussion_r137568572